[Bugfix][Frontend] Fix Issues Under High Load With `zeromq` Frontend

robertgshaw2-redhat opened the pull request:
### BUG REPO:
- On `v0.5.4` due to the mp frontend, we fail under heavy load:

- server script
```bash
vllm serve meta-llama/Meta-Llama-3-8B-Instruct
```

- client script:
```bash
python benchmarks/benchmark_serving.py --model meta-llama/Meta-Llama-3-8B-Instruct --dataset-name sonnet --sonnet-input-len 1024 --sonnet-output-len 128 --dataset-path benchmarks/sonnet.txt --port 8000 --num-prompts 10000 --request-rate 1000
```

The server dies when we get ~1000 active requests in the system

### SUMMARY:
- Multiprocessing OpenAI Frontend uses `icp` for the `zeromq`-based RPC connection. Each new request connects to this socket on the RPC Server side, meaning for N active requests, we will have N connections.
- `zmq.MAX_SOCKETS` defaults to 1023. If more than this number of sockets is opened we fail `Too many open files` error. This puts a hard cap on the number of active requests in vLLM and causes the bugs lists below.
- Setting `zmq.MAX_SOCKETS` to a high number resolves our issue. However, `lsof -U | wc -l` shows we are using `unix` sockets under the hood, the number of active requests will be limited by system `ulimit`. This is a bad UX and again puts a hard cap on the number of active requests running in vLLM.
- This PR introduces a `proxy` on the `RPCClient` side. The proxy creates a *single* `ipc` connection from the `RPCClient` to the `RPCServer`. Each API server thread handing a request connects to the proxy via  `inproc` protocol and the message is forwarded over the since `ipc` connection with proper `identity` tags. This means we only ever create 1 unix socket.
- As a bonus, this architecture is more performant by ~5% E2E, since we avoid so many syscalls to open unix sockets

### NOTES:
- Note: `zmq` has a variable called `zmq.MAX_SOCKETS` which applies at the `Context` level. This variable can be set up to a maximum of `zmq.SOCKET_LIMIT=65536`. This imposes a hard limit on the number of concurrent connections (see comments below for more details on this topic).

### PERFORMANCE:

Serving benchmark `H100` for `Llama-3-8B-Instruct` at `QPS=10`:

- server launch:
```bash
vllm serve meta-llama/Meta-Llama-3-8B-Instruct --enable-chunked-prefill --disable-log-requests
```

- benchmark script:
```bash
MODEL="meta-llama/Meta-Llama-3-8B-Instruct"
TOTAL_SECONDS=120
QPS_RATES=("10" "10" "10")
PORT=8000

for QPS in ${QPS_RATES[@]}; do
    NUM_PROMPTS=$((TOTAL_SECONDS * QPS))
    echo "===== RUNNING NUM_PROMPTS = $NUM_PROMPTS QPS = $QPS ====="

    python3 benchmarks/benchmark_serving.py \
        --model $MODEL \
        --dataset-name sonnet --sonnet-input-len 550 --sonnet-output-len 150 --dataset-path benchmarks/sonnet.txt \
        --num-prompts $NUM_PROMPTS --request-rate $QPS --port $PORT
done

```

- results:

| `branch` | `run 1` | `run 2` | `run 3` | **mean**
| - | - | - | - | - |
| `main` |  51.7 | 50.7 | 51.8 | **51.4** |
| `pr ` | 48.5 | 49.5 | 49.0 | **49.0**  |
| `--disable-frontend-multiprocessing` | 72.8 | 71.1 | 72.9 | **72.3** |



FIX 
- #7309
- #7290


### UPDATE 8/20 POST OFFLINE DISCUSSION

Per discussion offline with @njhill @simon-mo and @youkaichao

After much further investigation, we ran into a separate issue with the proxy design after enabling the proxy. Specifically, we encountered a situation where under high load, zeromq can DROP MESSAGES 😢 , due to a concept called high-water mark (https://zeromq.org/socket-api/#high-water-mark), which is designed to protect servers running zeromq from slow clients. Specifically:
- DEALER sockets go mute when the HWM is reached
- ROUTER sockets drop messages when the HWM is reached

This prevents issues where a server running ZMQ can run OOM due to a slow reciever which is not able to keep up with message passing. The HWM is defaulted to 1000. Its difficult to track down exactly the sequence of events that causes the message dropping under load, but the HWM seems to be the culprit

In our case, we disable the HWM 😨 . This is generally not advisable, however:
- We are in control of both the RPCServer + RPCClient
- Either the server OR the client dying is catastrophic for vllm + the whole system should restart in this case
- Any CPU OOM that occurs in the mp design will also occur in the non-mp design (+ we measured that overall memory usage is close)

Anyways:
- added a `test_load.py` to simulate load on the client and try to detect errors like this.
- added a `test_accuracy.py` to run an lm-eval-harness with concurrent requests via the openai API, which simulates real usage and makes sure we get the correct answer out of the server

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

Commit:
added proxy to limit use of uniz sockets

Commit:
Merge branch 'main' into fix-zmq-max-sockets

Commit:
comment

Commit:
use random inproc path

Commit:
format

Commit:
foamt

robertgshaw2-neuralmagic commented:
After further investigation, I using zmq sockets will impose a hard limit on the number of active request.

While it is possible to change the value of `ZMQ_MAX_SOCKET` on a `Context` by `Context` basis, the via `zmq_ctx_set`, `ZMQ_SOCKET_LIMIT`, which puts an upper bound on this. We are unable to change the value of `ZMQ_SOCKET_LIMIT` (see: https://libzmq.readthedocs.io/en/latest/zmq_ctx_get.html)

We are unable to change the value:
```python
import zmq
ctx = zmq.Context()
ctx.get(zmq.SOCKET_LIMIT)
# >>> 65535
ctx.set(zmq.SOCKET_LIMIT, 10000000)
ctx.get(zmq.SOCKET_LIMIT)
# >>> 65535
```

I believe this will put a hard limit on ZMQ at 65k concurrent connections.

This is obviously a lot of requests, but I am worried about an offline batch use case where someone sends 1M requests to the server at once. 

WDYT?

Commit:
format

njhill commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -21,12 +25,40 @@
+            if socket_from in events:
+                msg = await socket_from.recv_multipart()
+                await socket_to.send_multipart(msg)
+            elif socket_to in events:
```
In theory it could have both
```suggestion
            if socket_to in events:
```

robertgshaw2-redhat committed:
Update vllm/entrypoints/openai/rpc/client.py

Co-authored-by: Nick Hill <nickhill@us.ibm.com>

youkaichao commented:
@robertgshaw2-neuralmagic  thanks for the great work!

I think limiting the total number of pending requests make sense to me. Many web servers should have similar constraints.

As for offline batching inference, we don't need to surface this to users. We can accept arbitrary number of requests, but send part of them to the engine every time.

BTW, offline batch inference does not use this api server. So it should not be a concern there.

For the technical review, I would like to hand it over to @njhill  who has better expertise here.

robertgshaw2-neuralmagic commented:
Also note --- I tried running with `v0.5.3` and we seem to be dropping requests as is if we get >20k connections. I don't quite see why / what causes this without `zeromq`, but it means that our existing state did not support this either

robertgshaw2-neuralmagic commented on tests/tracing/test_tracing.py:
```diff
@@ -114,5 +114,5 @@ def test_traces(trace_service):
     assert attributes.get(SpanAttributes.LLM_LATENCY_E2E) == e2e_time
-    assert attributes.get(SpanAttributes.LLM_LATENCY_TIME_IN_SCHEDULER
-                          ) == metrics.scheduler_time
+    assert attributes.get(
```
make format happy

joerunde commented on vllm/entrypoints/openai/rpc/server.py:
```diff
@@ -177,11 +174,11 @@ async def run_server_loop(self):
         while True:
             # Wait for a request.
-            identity, message = await self.socket.recv_multipart()
+            identity, part2, message = await self.socket.recv_multipart()
```
For future readers it'd be nice to add a link to some zmq docs here or give this a descriptive name to say what `part2` is. From context here I'm guessing this is routing information for the client-side proxy?

robertgshaw2-neuralmagic replied:
Its related to the use of `ROUTER`, will do

robertgshaw2-neuralmagic commented:
> > Set --limit-concurrency in uvicorn to explicitly prohibit too many requests from being active in the server. This will give 503s if the Server is getting too many requests. (https://www.uvicorn.org/settings/#resource-limits)
> 
> @robertgshaw2-neuralmagic That would be great, the only issue I'd see is if the `/health` endpoint is used for a liveness probe, that could cause pod restarts under load. So I think we'd just want to also add a few words in the documentation recommending not to do that. (Looking at all the IBM deployment configs 😉 )

hmmmm - can you explain more?

joerunde commented:
> > > Set --limit-concurrency in uvicorn to explicitly prohibit too many requests from being active in the server. This will give 503s if the Server is getting too many requests. (https://www.uvicorn.org/settings/#resource-limits)
> > 
> > 
> > @robertgshaw2-neuralmagic That would be great, the only issue I'd see is if the `/health` endpoint is used for a liveness probe, that could cause pod restarts under load. So I think we'd just want to also add a few words in the documentation recommending not to do that. (Looking at all the IBM deployment configs 😉 )
> 
> hmmmm - can you explain more?

It's probably not a huge deal since vllm itself is a library and doesn't ship k8s configs but, in general readiness probes are used to back off traffic from pods experiencing issues accepting new requests, while liveness probes are used to kill pods that have crashed or experienced some unrecoverable error. If we were to set `--limit-concurrency`, then the `/health` endpoint would also start returning 503 when the pod is busy processing the maximum number of requests. That would be great to use as a readiness probe, k8s would automatically route traffic away while the pod is busy. But, if somebody accidentally used the `/health` endpoint as a liveness probe, then k8s would reboot the pod when it's busy. This would then likely cause that traffic to hit other pods, causing them to reach the maximum request limit and restart, and you'd end up in a cascading failure.

I think we'd want to suggest something different to use as a liveness probe in that case, maybe something as simple as checking that the frontend and backend processes are running.

robertgshaw2-neuralmagic commented:
> > > > Set --limit-concurrency in uvicorn to explicitly prohibit too many requests from being active in the server. This will give 503s if the Server is getting too many requests. (https://www.uvicorn.org/settings/#resource-limits)
> > > 
> > > 
> > > @robertgshaw2-neuralmagic That would be great, the only issue I'd see is if the `/health` endpoint is used for a liveness probe, that could cause pod restarts under load. So I think we'd just want to also add a few words in the documentation recommending not to do that. (Looking at all the IBM deployment configs 😉 )
> > 
> > 
> > hmmmm - can you explain more?
> 
> It's probably not a huge deal since vllm itself is a library and doesn't ship k8s configs but, in general readiness probes are used to back off traffic from pods experiencing issues accepting new requests, while liveness probes are used to kill pods that have crashed or experienced some unrecoverable error. If we were to set `--limit-concurrency`, then the `/health` endpoint would also start returning 503 when the pod is busy processing the maximum number of requests. That would be great to use as a readiness probe, k8s would automatically route traffic away while the pod is busy. But, if somebody accidentally used the `/health` endpoint as a liveness probe, then k8s would reboot the pod when it's busy. This would then likely cause that traffic to hit other pods, causing them to reach the maximum request limit and restart, and you'd end up in a cascading failure.
> 
> I think we'd want to suggest something different to use as a liveness probe in that case, maybe something as simple as checking that the frontend and backend processes are running.

Thanks - this is clear

Commit:
cleaning

robertgshaw2-redhat committed:
Merge branch 'main' into fix-zmq-max-sockets

Commit:
Merge branch 'fix-zmq-max-sockets' of https://github.com/neuralmagic/vllm into fix-zmq-max-sockets

Commit:
remove logging

Commit:
add info message re: concurrency

Commit:
update comment

Commit:
update

Commit:
format

Commit:
reorder

Commit:
reverT

Commit:
fix

Commit:
fix

Commit:
fix abort logic

Commit:
reduce LOC change

Commit:
cleanup

Commit:
cleanup

robertgshaw2-neuralmagic commented on vllm/entrypoints/openai/rpc/server.py:
```diff
@@ -41,78 +42,66 @@ def cleanup(self):
-
-    async def generate(self, identity, generate_request: RPCGenerateRequest):
+            await self.socket.send_multipart(
+                [rpc_id, client_id,
```
note: if there is an exception calling `abort`, we should propagate this to the client

joerunde replied:
👍 , makes sense

Commit:
format

Commit:
fix client

Commit:
revert unneccessary change

Commit:
revert startup probe changes to separate PR

Commit:
stash

Commit:
Merge branch 'main' into fix-zmq-max-sockets

Commit:
stash draining

Commit:
update

Commit:
stash

Commit:
convert RPCServer to use DEALER

Commit:
stash

Commit:
fix

Commit:
cleaning

robertgshaw2-neuralmagic commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -8,31 +10,136 @@
+
+        # IPC connection to RPC Server (uses unix sockets).
+        self.to_rcp_server = self.context.socket(zmq.constants.DEALER)
+        self.to_rcp_server.set_hwm(0)
```
@njhill - this connects to the RPCServer via IPC ... setting the hwm here seems to make things work properly

Commit:
stash

Commit:
remove awk

njhill reviewed (commented):
Thanks for all of the hard work on this @robertgshaw2-neuralmagic!

Wondering if we should _also_ disable the hwm on the inproc sockets? so we can guarantee no messages are silently dropped (even though disabling it for the client->server ipc effectively avoids it)

Also formatting needs fixing

robertgshaw2-neuralmagic commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -93,30 +189,46 @@ async def _send_get_data_rpc_request(self, request: RPCUtilityRequest,
             if expected_type == LoRAConfig and data is None:
                 pass
+            elif isinstance(data, Exception):
+                logger.warning(error_message)
```
I wanted to raise the error that actually arose on the Engine side

robertgshaw2-neuralmagic replied:
could be a different PR

Commit:
nits

Commit:
format

Commit:
format

Commit:
nit

Commit:
change

Commit:
clean

robertgshaw2-neuralmagic commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -93,30 +189,46 @@ async def _send_get_data_rpc_request(self, request: RPCUtilityRequest,
 
         if not isinstance(response, str) or response != VLLM_RPC_SUCCESS_STR:
+            if isinstance(response, Exception):
+                logger.warning(error_message)
```
I wanted to raise the error that actually arose on the Engine side rather than a madeup ValueError

robertgshaw2-neuralmagic replied:
could be a different PR

robertgshaw2-redhat committed:
Update vllm/entrypoints/openai/rpc/server.py

Co-authored-by: Nick Hill <nickhill@us.ibm.com>

Commit:
format

njhill commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -242,47 +356,63 @@ async def generate(
-                            await self.check_health()
-                        except Exception:
-                            self._errored = True
+
```
sorry I think I commented on the wrong line before, I meant to suggest adding the `finished = True` here rather than within the `except` block below.

robertgshaw2-neuralmagic replied:
but don't we want `finished = False`, such that we call `abort`?

robertgshaw2-neuralmagic replied:
+ then we only set `finished=True` in the `except` block such that we don't call self.abort() below

njhill commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -242,47 +356,63 @@ async def generate(
+                        request_output = cloudpickle.loads(message)
+
+                        # If aborted, CancelledError is injected.
+                        if request_output == asyncio.exceptions.CancelledError:
```
I noticed that you didn't change this per my prior comment... probably because the current code is what works :)

It _shouldn't_ work though ... I traced this to a bug which I think I introduced myself recently. Opening a PR to fix that now.

njhill replied:
https://github.com/vllm-project/vllm/pull/7698 .. just adding a test

njhill reviewed (commented):
@robertgshaw2-neuralmagic I guess there's one of those yapf/ruff conflicts which requires brute-force disabling of yapf

njhill committed:
[BugFix] Raise all exception variations in async generator

The async generator returned by AsyncLLMEngine generate() uses a queue of either request outputs and/or exceptions.

But the logic consuming requests from the queue only raises instances of Exceptions, not Exception classes (which are also raisable in python). This means that the CancelledError class instance we add to the queue elsewhere currently isn't being raised from the generator as it should be.

Commit:
cleanup abort logic

Commit:
nit

robertgshaw2-neuralmagic commented:
/ready

Commit:
added load test

Commit:
update load test

njhill committed:
Fix possible premature generator completion; add tests

Commit:
updated

Commit:
format

Commit:
updated

Commit:
revert suurious change

Commit:
convert to even smaller model

Commit:
20k requests

robertgshaw2-neuralmagic commented on vllm/entrypoints/openai/api_server.py:
```diff
@@ -131,6 +131,9 @@ async def build_async_engine_client(args) -> AsyncIterator[AsyncEngineClient]:
         logger.info("Multiprocessing frontend to use %s for RPC Path.",
                     rpc_path)
 
+        # Build RPCClient, which conforms to AsyncEngineClient Protocol.
```
moved first, since we `bind` here
- note: this shouldn't matter, but just being safe

Commit:
convert to 10k requests

Commit:
clean up closing logic

Commit:
use constant

Commit:
fix bad cleanup

Commit:
remove useless argument

Commit:
up to 20k requests

Commit:
revert to 10k requests

Commit:
revert suprious argument

Commit:
revert to 20k

Commit:
format

Commit:
format

Commit:
added test accuracy

Commit:
format

Commit:
updated test pipeline

Commit:
fix lm eval

robertgshaw2-neuralmagic commented on .buildkite/test-pipeline.yaml:
```diff
@@ -86,6 +86,7 @@ steps:
   - vllm/
   commands:
   - pip install -e ./plugins/vllm_add_dummy_model
+  - pip install git+https://github.com/EleutherAI/lm-evaluation-harness.git@a4987bba6e9e9b3f22bd3a6c1ecf0abd04fd5622#egg=lm_eval[api]
```
need to install from source since `local-completions` is not yet in release of `lm_eval`

robertgshaw2-neuralmagic commented on vllm/entrypoints/openai/rpc/client.py:
```diff
@@ -8,31 +10,137 @@
+        self.from_api_server.bind(INPROC_PROXY_PATH)
+
+        # Asyncio background task for the proxy.
+        self.proxy_task = asyncio.create_task(
```
@njhill does this need to be explicitly canceled somewhere?

(e.g. in `close()`)

njhill replied:
@robertgshaw2-neuralmagic yes, we should cancel it there.

robertgshaw2-neuralmagic commented on tests/entrypoints/openai/test_load.py:
```diff
@@ -0,0 +1,105 @@
+"""
+This file tests significant load on the vLLM server.
```
cc @simon-mo --- this test takes ~3 minutes on H100

Will likely take >10 min on L4 ... are you okay with this?

Commit:
cleanup

Commit:
updated

Commit:
Merge branch 'main' into fix-zmq-max-sockets

Commit:
added sleep time

Commit:
actually sleep

Commit:
formatting

robertgshaw2-neuralmagic commented:
`test_load.py` is failing in the CI due to `Too Many Open Files` ---> It seems like this happens at the very beginning which is odd ... looking into it

Commit:
format

Commit:
mypy

Commit:
mypy

Commit:
format

Commit:
remove test load

Commit:
stash

Commit:
Merge branch 'fix-zmq-max-sockets' of https://github.com/neuralmagic/vllm into fix-zmq-max-sockets

Commit:
Merge branch 'fix-raise-cancelled' into fix-zmq-max-sockets

robertgshaw2-neuralmagic commented:
Should be merged after: https://github.com/vllm-project/vllm/pull/7698

njhill reviewed (approved):
Thanks @robertgshaw2-neuralmagic!
